core/src/main/java/io/grpc/transport/MessageDeframer.java

```diff
@@ -174,6 +303,117 @@ public void deframe(ReadableBuffer data, boolean endOfStream) {
+    long unreturnedBytes = bytesReceived - bytesReturned;
+    long bytesToReturnImmediate = 0;
+    try {
+      // Process the uncompressed bytes.
+      CompositeReadableBuffer buffer = readRequiredBytes();
+      while (buffer != null) {
+        bytesConsumed += buffer.readableBytes();
+        bytesToReturnImmediate += returnBytes(true, queue);
+        switch (state) {
+          case HEADER:
+            processHeader(buffer);
+            break;
+          case BODY:
+            // Read the body and enqueue the message.
+            processBody(buffer, queue);
+            readBody = true;
+            break;
+          default:
+            throw new AssertionError("Invalid state: " + state);
+        }
+        buffer = readRequiredBytes();
+      }
+      if (endOfStream) {
+        if (state == State.BODY || unprocessed.readableBytes() > 0) {
+          // We've received the entire stream and have data available but we don't have
+          // enough to read the next frame ... this is bad.
+          throw Status.INTERNAL.withDescription("Encountered end-of-stream mid-frame")
+              .asRuntimeException();
+        } else {
+          // Can't notify stream yet as there were not enough message requests available
+          // to consume all the messages. Instead we have to enqueue signalling EOS on the
+          // stream.
+          if (transportCount.get() < 0) {
+            queuedEos = true;
+            queue.add(END_OF_STREAM);
+          }
+        }
+      }
+      // Can immediately return all bytes that have not been returned via the queue at this point.
+      bytesToReturnImmediate += returnBytes(false, queue);
+      if (bytesToReturnImmediate > 0) {
+        listener.bytesRead((int) bytesToReturnImmediate);
+      }
+    } finally {
+      queueRef.getAndSet(queue);
+    }
+
+    // We are stalled when there is insufficient data to read another frame.
+    // This allows delivering errors as soon as the buffered input has been consumed,
+    // independent of whether the application has requested another message.
+    boolean stalled = unprocessed.readableBytes() == 0 || readBody;
+
+    // Must notify if we've read a body or we had to return flow-control via the queue.
+    if (readBody || unreturnedBytes > bytesToReturnImmediate) {
+      // Notify the listener that messages may be available. It's possible that the messages are
+      // already being consumed in another thread that is racing with this one and that the
+      // listener will have nothing to process as a result of this call but we must do
+      // it to guarantee that all payloads are consumed.
+      listener.messagesAvailable(this);
+    }
+
+    // Can notify the listener immediately instead of via the queue as there we enough
+    // request tokens visible to the transport when it enqueued messages to know that
+    // the queue can be fully drained by the application.
+    if (endOfStream && !queuedEos) {
+      listener.endOfStream();
```

Comment:
How much of an optimization is this? Is it logically necessary or will it work if we always enqueue the EOS frame?